src/workload.py

```diff
@@ -0,0 +1,287 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""MySQL Router workload"""
+
+import configparser
+import dataclasses
+import logging
+import pathlib
+import socket
+import string
+import typing
+
+import ops
+
+import mysql_shell
+
+if typing.TYPE_CHECKING:
+    import charm
+    import relations.database_requires
+
+logger = logging.getLogger(__name__)
+
+
+@dataclasses.dataclass(kw_only=True)
+class Workload:
+    """MySQL Router workload"""
+
+    _container: ops.Container
+
+    CONTAINER_NAME = "mysql-router"
+    _SERVICE_NAME = "mysql_router"
+    _UNIX_USERNAME = "mysql"
+    _ROUTER_CONFIG_DIRECTORY = pathlib.Path("/etc/mysqlrouter")
+    _ROUTER_CONFIG_FILE = "mysqlrouter.conf"
+    _TLS_CONFIG_FILE = "tls.conf"
+
+    @property
+    def container_ready(self) -> bool:
+        """Whether container is ready"""
+        return self._container.can_connect()
+
+    @property
+    def _enabled(self) -> bool:
+        """Service status"""
+        service = self._container.get_services(self._SERVICE_NAME).get(self._SERVICE_NAME)
+        if service is None:
+            return False
+        return service.startup == ops.pebble.ServiceStartup.ENABLED
+
+    @property
+    def version(self) -> str:
+        """MySQL Router version"""
+        process = self._container.exec(["mysqlrouter", "--version"])
+        raw_version, _ = process.wait_output()
+        for version in raw_version.split():
+            if version.startswith("8"):
+                return version
+        return ""
+
+    def _update_layer(self, *, enabled: bool, tls: bool = None) -> None:
+        """Update and restart services.
+
+        Args:
+            enabled: Whether MySQL Router service is enabled
+            tls: Whether TLS is enabled. Required if enabled=True
+        """
+        if enabled:
```

Comment:
what happens if there is no relation between tls-operator and mysql-router? why must tls be specified when enabled=True?

Reply:
Is this checking something else? What you described seems like:

```
if not enabled:
    assert tls is None
```